### PR TITLE
[*] Update interface and abstract class viewmodel

### PR DESCRIPTION
**Pull request by Yuhan-Ut**
make an abstract class for Viewmodel, so the update interface (and it's abstract method)can accept all viewmodel, but not only bill_update
also, sorted usecases files under the package called usecases

### CODE REVIEW

File: src/main/java/billgates/usecases/alter_entry/AlterEntryOutputPort.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package billgates.usecases.alter_entry;
+
+
+public interface AlterEntryOutputPort {
+    void updateBill(AlterEntryResponseModel model);
```
Comment (Yunshan-522) [inline src/main/java/billgates/usecases/alter_entry/AlterEntryOutputPort.java:5]: I guess this should be: viod alterEntry(AlterEntryResponseModel model);

File: src/main/java/billgates/usecases/alter_entry/AlterEntryController.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import billgates.database.QueryEntryData;
+
+public class AlterEntryController {
+    private final AlterEntryInputPort useCase;
+
+    public AlterEntryController(AlterEntryInputPort useCase) {
+        this.useCase = useCase;
+    }
+
+    public void update(int billId, QueryEntryData entry) {
```
Comment (Yunshan-522) [inline src/main/java/billgates/usecases/alter_entry/AlterEntryController.java:12]: I guess this should be: public viod alter(int billId, QueryEntryData entry) {

File: src/main/java/billgates/usecases/alter_entry/AlterEntryPresenter.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+public class AlterEntryPresenter implements AlterEntryOutputPort {
+
+    private final Updatable view;
+
+    public AlterEntryPresenter(Updatable view) {
+        this.view = view;
+    }
+
+    @Override
+    public void updateBill(AlterEntryResponseModel model) {
```
Comment (Yunshan-522) [inline src/main/java/billgates/usecases/alter_entry/AlterEntryPresenter.java:17]: I guess this should be: public void alterEntry(AlterEntryResponseModel model) {

File: src/main/java/billgates/usecases/alter_entry/AlterEntryPresenter.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        List<List<Object>> entries = model.getEntries();
+        // TODO: We should format the column in a more intuitive way.
+        AlterEntryViewModel viewModel = new AlterEntryViewModel(entries);
+        this.view.update(viewModel);
```
Comment (Yunshan-522) [inline src/main/java/billgates/usecases/alter_entry/AlterEntryPresenter.java:21]: Same there: this.view.alter(viewModel);

File: src/main/java/billgates/usecases/alter_entry/AlterEntryUseCase.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+
+    @Override
+    public void alterEntry(int billId, QueryEntryData entry) {
+        gateway.modifyEntry(billId, entry);
+        List<Entry> result = this.gateway.getBillData(User.getInstance().getCurrentBillID()).getEntries()
+                .stream().map(d -> d.toEntryBuilder().build()).toList();
+        System.out.println(result); // TODO remove this
+        this.presenter.updateBill(new AlterEntryResponseModel(result.stream().map(Entry::toObjects).toList()));
```
Comment (Yunshan-522) [inline src/main/java/billgates/usecases/alter_entry/AlterEntryUseCase.java:26]: Same there: this.presenter.alterEntry(

**Review by Yunshan-522 — COMMENTED** [review]
I think it's a great job, especially the idea of linking all viewmodel is excellent! However, I think some functions should be named alter() or alterEntry(). Since these functions all alter the entry, maybe it would be easier to understand if they were renamed alter or alterEntry.

**Comment by xgao28** [comment]
Also, this pull request consists of commits from others, so you may want to open another pull request.

**Comment by Yuhan-Ut** [comment]
> Also, this pull request consists of commits from others, so you may want to open another pull request.

Do you know how to avoid this? I did not noticed there are scott's commit here. How to start a pr that only my commit?

**Comment by xgao28** [comment]
> > Also, this pull request consists of commits from others, so you may want to open another pull request.
> 
> Do you know how to avoid this? I did not noticed there are scott's commit here. How to start a pr that only my commit?

Creating a new branch that's up-to-date, and coding in the new branch may work?

**Comment by Lei-Tin** [comment]
I will not review this Pull Request at the moment, as it is not really urgent to be implemented. Furthermore, with Scott's `ViewModel` & `Presenter` Demo, I believe it is enough to test out these use cases. 

Also, I believe by implementing the use cases first and completely, we can then decide on display and rendering issues later on. 

But, nice start!

**Comment by xgao28** [comment]
I reviewed the general idea of this pull request.
Please refer my thoughts in this discussion. #37

**Pull request closed without merging**